### PR TITLE
Dokka. API reference

### PR DESCRIPTION
**Pull request by aSemy**
This PR updates the build config to generates Dokka HTML API code reference for all subprojects.

To generate the API docs: 

1. Run `./gradlew :docs:dokkaGen`
2. Open `docs/build/dokka/html/index.html` in a browser.

Part of #2228.

The next step is publishing the API reference to a website (e.g. GitHub Pages).

### Screenshots

Here are some example screenshots.

* The home page shows a list of all modules, each with a brief description (from the README.md in each subproject)
  
  image

* When viewing a module, all content from the module's README.md is available
  
  image

* Code and KDoc are easily viewable and searchable
  image

### Summary 

- Update subproject READMEs to use as Dokka Module files (add `# Module` header)
- Generate aggregated Dokka HTML in the `:docs` subproject
- Remove 'archived' from headings - Dokka needs the heading to match the module name _exactly_. Instead, use admonitions. But... Dokka doesn't support them. So instead, find/replace the admonitions with emojis.

### TODO

- [x] Add logo
- [x] Add source link
- [x] Add homepage link

### Reasons for using Dokkatoo over the existing Dokka Gradle Plugin.

- Dokkatoo compatible with the Gradle API and follows best practices, resulting in better performance. DGP is not compatible, resulting in worse build performance and more Gradle warnings.
- kotlin-wrappers has a `:docs` subproject. It makes sense to aggregate the API docs into this subproject. DGP cannot create an aggregated HTML doc into a subproject, while Dokkatoo can.

### CODE REVIEW

File: buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
+    id("dev.adamko.dokkatoo-html")
 }
 
 kotlin {
     js {
         moduleName = project.name
 
         when (project.jsPlatform) {
-            JsPlatform.WEB -> {
+            JsPlatform.WEB     -> {
```
Comment (aSemy) [inline buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts:17]:
```suggestion
            JsPlatform.WEB -> {
```

File: buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts [hunk 2]
```diff
@@ -21,7 +23,7 @@ kotlin {
                 browser()
             }
 
-            JsPlatform.NODE -> {
+            JsPlatform.NODE    -> {
```
Comment (aSemy) [inline buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts:26]:
```suggestion
            JsPlatform.NODE -> {
```

File: gradle.properties [hunk 3]
```diff
@@ -2,9 +2,11 @@ group=org.jetbrains.kotlin-wrappers
 version.build=pre.754
 
 org.gradle.jvmargs=-Xms2g -Xmx4g
+org.gradle.caching=true
 org.gradle.parallel=true
 org.gradle.caching=true
 
+
```
Comment (aSemy) [inline gradle.properties:9]:
```suggestion
```

File: buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts [hunk 4]
```diff
@@ -32,6 +34,81 @@ kotlin {
     }
 }
 
+val prepareReadmeForDokka by tasks.registering {
```
Comment (turansky) [inline buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts:37]: Could we move it in separate plugin like `dokka-conventions`?

**Review by turansky — CHANGES_REQUESTED** [review]

File: kotlin-styled-next/build.gradle.kts [hunk 5]
```diff
@@ -33,22 +35,27 @@ dependencies {
...
-            "CssBuildersInject",
-            "AddDuplicateCss",
-            "DataTypeOperations",
-            "InjectCssNPlusOne"
-        )
-        fileNames.forEach { test ->
-            val report = layout.buildDirectory
-                .file("reports/tests/test/classes/benchmark.$test.html")
-                .get().asFile.readText()
+    // implicit dependency to benchmark classes names
```
Comment (turansky) [inline kotlin-styled-next/build.gradle.kts:38]: Could we move this change in separate PR?

**Review by turansky — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts (hunk 4 above)
Comment (aSemy) [reply buildSrc/src/main/kotlin/kotlin-conventions.gradle.kts]: Good idea, done!

File: buildSrc/src/main/kotlin/dokka-conventions.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import gradle.kotlin.dsl.accessors._74857f9dd946e843b9fa7c149d123d57.dokkatoo
```
Comment (turansky) [inline buildSrc/src/main/kotlin/dokka-conventions.gradle.kts:1]: Redundant?

**Review by turansky — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/dokka-conventions.gradle.kts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import gradle.kotlin.dsl.accessors._74857f9dd946e843b9fa7c149d123d57.html
```
Comment (aSemy) [inline buildSrc/src/main/kotlin/dokka-conventions.gradle.kts:2]:
```suggestion
```

File: buildSrc/src/main/kotlin/dokka-conventions.gradle.kts (hunk 6 above)
Comment (aSemy) [reply buildSrc/src/main/kotlin/dokka-conventions.gradle.kts]: yes, IntelliJ sometimes adds this when copy/pasting between build scripts.

File: kotlin-styled-next/build.gradle.kts (hunk 5 above)
Comment (aSemy) [reply kotlin-styled-next/build.gradle.kts:38]: Sure thing https://github.com/JetBrains/kotlin-wrappers/pull/2380

File: kotlin-wrappers-bom/build.gradle.kts [hunk 8]
```diff
@@ -1,12 +1,17 @@
...
+    dependencyConstraints.addAllLater(
+        kotlinWrapperSubprojects.bomDependencies.map { subproject ->
+            logger.info("[$path] adding ${subproject.size} subprojects to BOM: $subproject")
+            subproject.sorted().map { coord ->
+                project.dependencies.constraints.create(project(coord))
+            }
         }
-    }
+    )
 }
```
Comment (adam-enko) [inline kotlin-wrappers-bom/build.gradle.kts:17]: TODO remove kotlinWrapperSubprojects service. It will be reviewed in a separate PR https://github.com/JetBrains/kotlin-wrappers/pull/2379/

File: gradle.properties [hunk 9]
```diff
@@ -2,8 +2,8 @@ group=org.jetbrains.kotlin-wrappers
 version.build=pre.790
 
 org.gradle.jvmargs=-Xms2g -Xmx4g
-org.gradle.parallel=true
 org.gradle.caching=true
+org.gradle.parallel=true
```
Comment (turansky) [inline gradle.properties:6]: Redundant change?

**Review by turansky — CHANGES_REQUESTED** [review]

**Comment by turansky** [comment]
@aSemy Thank you!
